### PR TITLE
Add bh

### PR DESCRIPTION
**Pull request by TheCoolKidsTable**

### CODE REVIEW

File: module/vision/BallDetector/src/BallDetector.cpp [hunk 1]
```diff
@@ -157,13 +157,22 @@ namespace module::vision {
...
                         b.screen_angular = cartesianToSpherical(axis).tail<2>();
                         b.angular_size   = Eigen::Vector2f::Constant(std::acos(radius));
 
+                        //Convert to rBCc
+                        // float r = 1/distance;
+                        // const Eigen::Vector3d srBCc =  b.measurements.back().srBCc;
+                        // rBCc.x() = r*std::cos(srBCc.z())*std::sin(srBCc.y());
+                        // rBCc.y() = r*std::sin(srBCc.z())*std::sin(srBCc.y());
+                        // rBCc.z() = r*std::cos(srBCc.y());
+                        // log(rBCc.x(),rBCc.y(),rBCc.z());
```
Comment (tayloryoung6396) [inline module/vision/BallDetector/src/BallDetector.cpp:174]: WTF

File: module/input/GameController/data/config/GameController.yaml [hunk 2]
```diff
@@ -3,4 +3,4 @@ log_level: WARN
 player_id: 1
 receive_port: 3838
 send_port: 3939
-udp_filter_address: "" # If set, will only accept GameController packets from this IP address
+udp_filter_address: "10.1.0.103" # If set, will only accept GameController packets from this IP address
```
Comment (tayloryoung6396) [inline module/input/GameController/data/config/GameController.yaml]: Remove this shit!

File: module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp [hunk 3]
```diff
@@ -472,11 +516,12 @@ namespace module::behaviour::strategy {
...
-        emit(std::make_unique<MotionCommand>(utility::behaviour::WalkToState(
-            utility::math::transform::lookAt(position,
-                                             Eigen::Vector2d(fieldDescription.dimensions.field_length * 0.5, 0.0)))));
-    }
+    // void SoccerStrategy::walkTo(const FieldDescription& fieldDescription, const Eigen::Vector2d& position) {
+    //     log<NUClear::DEBUG>("walkTo() position");
+    //     emit(std::make_unique<MotionCommand>(utility::behaviour::WalkToState(
+    //     utility::math::transform::lookAt(position,
+    //                                         Eigen::Vector2d(fieldDescription.dimensions.field_length * 0.5, 0.0)))));
+    // }
```
Comment (tayloryoung6396) [inline module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp:524]: This should go later...

**Review by tayloryoung6396 — COMMENTED** [review]
LGTM

File: module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp [hunk 4]
```diff
@@ -324,21 +348,24 @@ namespace module::behaviour::strategy {
...
         standStill();
         currentState = Behaviour::State::FINISHED;
     }
 
     // ********************NORMAL GAMEMODE STATES********************************
     void SoccerStrategy::normalInitial() {
+        log<NUClear::DEBUG>("normalInitial()");
         standStill();
-        find({FieldTarget(FieldTarget::Target::SELF)});
+        // find({FieldTarget(FieldTarget::Target::SELF)});
```
Comment (KipHamiltons) [inline module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp:368]: seems strange? also, commented code

File: module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp [hunk 5]
```diff
@@ -350,71 +377,85 @@ namespace module::behaviour::strategy {
...
+        // walkTo(fieldDescription, walkTarget);
+
+        if (!startedWalkingToReady) {
+            startedWalkingToReadyAt = NUClear::clock::now();
+            startedWalkingToReady = true;
         }
-        else {
-            walkTo(fieldDescription, cfg_.start_position_defensive);
+
+        if (NUClear::clock::now() - startedWalkingToReadyAt < std::chrono::milliseconds(17*1000)) {
```
Comment (KipHamiltons) [inline module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp:390]: magic number. any rationale behind it?

File: module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp [hunk 6]
```diff
@@ -350,71 +377,85 @@ namespace module::behaviour::strategy {
...
+            emit(std::make_unique<MotionCommand>(utility::behaviour::WalkToReady()));
+        } else {
+            log("10 seconds up, stopping walk");
+            standStill();
         }
-        find({FieldTarget(FieldTarget::Target::SELF)});
+
+        // Self localise while we're walking on in READY
+        // find({FieldTarget(FieldTarget::Target::SELF)});
```
Comment (KipHamiltons) [inline module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp:398]: wondering the general idea behind commenting all of these

File: module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp [hunk 7]
```diff
@@ -350,71 +377,85 @@ namespace module::behaviour::strategy {
...
+
         currentState = Behaviour::State::READY;
     }
 
     void SoccerStrategy::normalSet() {
+        log<NUClear::DEBUG>("normalSet()");
         standStill();
-        find({FieldTarget(FieldTarget::Target::BALL)});
+        // find({FieldTarget(FieldTarget::Target::BALL)});
```
Comment (KipHamiltons) [inline module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp:406]: If you're in normal set, shouldn't you be finding yourself? - someone cool, probably

File: module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp [hunk 8]
```diff
@@ -350,71 +377,85 @@ namespace module::behaviour::strategy {
...
     }
 
     void SoccerStrategy::normalPlaying(const Field& field, const Ball& ball, const FieldDescription& fieldDescription) {
+        log<NUClear::DEBUG>("normalPlaying()");
         if (penalised() && !cfg_.forcePlaying) {  // penalised
-            standStill();
-            find({FieldTarget(FieldTarget::Target::SELF)});
+            // standStill();
+            log("we think we are penalised");
+            find({FieldTarget(FieldTarget::Target::BALL)});
```
Comment (KipHamiltons) [inline module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp:416]: shouldn't you localise yourself, when you're penalised? you can teleport to the other side of the field, so localising self might be important?? @tayloryoung6396  thoughts?

File: module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp [hunk 9]
```diff
@@ -350,71 +377,85 @@ namespace module::behaviour::strategy {
...
             currentState = Behaviour::State::PENALISED;
+            return;
```
Comment (KipHamiltons) [inline module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp:418]: don't need it

File: module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp [hunk 10]
```diff
@@ -350,71 +377,85 @@ namespace module::behaviour::strategy {
...
         }
         else if (cfg_.is_goalie) {  // goalie
             find({FieldTarget(FieldTarget::Target::BALL)});
             goalieWalk(field, ball);
             currentState = Behaviour::State::GOALIE_WALK;
         }
+
+        if (NUClear::clock::now() - ballLastMeasured < cfg_.ball_last_seen_max_time && std::abs(rBTt.x()) < 10 && std::abs(rBTt.y()) < 10 && std::abs(rBTt.z()) < 10) {
```
Comment (KipHamiltons) [inline module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp:426]: why 10. ideal world adds a comment

File: module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp [hunk 11]
```diff
@@ -350,71 +377,85 @@ namespace module::behaviour::strategy {
...
+            emit(std::make_unique<MotionCommand>(utility::behaviour::RotateOnSpot()));
+            find({FieldTarget(FieldTarget::Target::BALL)});
         }
     }
 
     void SoccerStrategy::normalFinished() {
+        log<NUClear::DEBUG>("normalFinished()");
         standStill();
-        find({FieldTarget(FieldTarget::Target::SELF)});
+        // find({FieldTarget(FieldTarget::Target::SELF)});
```
Comment (KipHamiltons) [inline module/behaviour/strategy/SoccerStrategy/src/SoccerStrategy.cpp:446]: these are kinda whack. these commented out bits. idk :shrimp:

File: module/vision/BallDetector/src/BallDetector.cpp (hunk 1 above)
Comment (KipHamiltons) [reply module/vision/BallDetector/src/BallDetector.cpp:174]: "we do this somewhere else"

**Review by KipHamiltons — COMMENTED** [review]
yaaaay 7 am review yaaaay :tada: :taco: :shrimp: :fried_egg:

**Review by KipHamiltons — APPROVED** [review]
This is disgustingly hacky. yikes. But, I am taking you on your word that it's an improvement on the stuff in master currently. If that's true, then ship it baby lets get this merged yewwww.
I can't see anything that is obviously (unintentionally) broken (localisation is intentionally broken, so that doesn't count lol). sure.

File: module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp [hunk 12]
```diff
@@ -164,76 +178,74 @@ namespace module::behaviour::planning {
                     emit(std::move(command));
                     emit(std::make_unique<ActionPriorities>(ActionPriorities{subsumptionId, {40, 11}}));
                     return;
+                }   else if (latestCommand.type == message::behaviour::MotionCommand::Type::WALK_TO_READY_POSITION) {
+                    // log("Walk to ready");
+                    std::unique_ptr<WalkCommand> command =
+                        std::make_unique<WalkCommand>(subsumptionId,
+                                                  Eigen::Vector3d(walkToReadySpeedX, walkToReadySpeedY, walkToReadyRotation));
+                    emit(std::move(command));
+                    emit(std::make_unique<ActionPriorities>(ActionPriorities{subsumptionId, {40, 11}}));
```
[the remark this answers is not in the document]
Comment (Bidski) [reply module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp:187]: Changes priorities for each item in the limb set

[the remark this answers is not in the document]
Comment (Bidski) [reply module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp:187]: The magic is some relative importance on how much you want those limbs

File: module/platform/Webots/src/Webots.cpp [hunk 13]
```diff
@@ -792,8 +792,13 @@ namespace module::platform {
...
             image->data           = camera.image;
 
-            image->id        = camera_context[camera.name].id;
-            image->timestamp = NUClear::clock::time_point(std::chrono::nanoseconds(sensor_measurements.time));
+            image->id = camera_context[camera.name].id;
+            // TODO unhack
+            // Webots clock and custom clock get out of sync very quickly. HeadBehaviour requires accurate timestamps
+            // on images to ensure it doesn't get stuck fixated on the ball. Safer at this point to save the timestamp
+            // using NUClear clock, however the unsync issue should be investigated
+            image->timestamp = NUClear::clock::now();
```
Comment (Bidski) [inline module/platform/Webots/src/Webots.cpp:800]: Setting it to `NUClear::clock::now()` should actually be preferred to what we had before anyway
```suggestion
            image->timestamp = NUClear::clock::now();
```

**Review by Bidski — CHANGES_REQUESTED** [review]

**Comment by KipHamiltons** [comment]
What's happening with this PR @TheCoolKidsTable? Does it have stuff you want to keep/merge?

**Pull request closed without merging**